sklearn/tree/_utils.pyx

```diff
@@ -25,17 +24,19 @@ np.import_array()
 # Helper functions
 # =============================================================================
 
-cdef realloc_ptr safe_realloc(realloc_ptr* p, size_t nelems) except *:
+cdef realloc_ptr safe_realloc(realloc_ptr* p, size_t nelems) nogil except *:
```

Comment:
I tried casting `NULL` to whatever type `realloc_ptr` was like

```py
%%cython

ctypedef fused realloc_ptr:
    (unsigned char*)
    (size_t*)
    
cdef realloc_ptr safe_realloc(realloc_ptr* p) nogil except (<realloc_ptr> NULL):
    with gil:
        raise MemoryError("I ain't gonna work...")
    return NULL


cpdef fn():
    cdef size_t* a
    safe_realloc(&a)
    
fn()
```

But cython barfs saying --

```
....
....
Exception: This may never happen, please report a bug
```

:P